dcalc/DmpCeff.cc

```diff
@@ -1766,30 +1766,27 @@ DmpError::DmpError(const char *what) :
   //printf("DmpError %s\n", what);
 }
 
-// This saves about 2.5% in overall run time on designs with SPEF.
-// https://codingforspeed.com/using-faster-exponential-approximation
+// Fast approximate exponential. See
+// https://www.schraudolph.org/pubs/Schraudolph99.pdf
 static double
 exp2(double x)
 {
-  if (x < -12.0)
-    // exp(-12) = 6.1e-6
-    return 0.0;
-  else {
-    double y = 1.0 + x / 4096.0;
-    y *= y;
-    y *= y;
-    y *= y;
-    y *= y;
-    y *= y;
-    y *= y;
-    y *= y;
-    y *= y;
-    y *= y;
-    y *= y;
-    y *= y;
-    y *= y;
-    return y;
+  if (std::abs(x) > 707.703272) {
+    // For arguments with magnitude greater than ln(DBL_MAX/8) ~= 707.703272,
+    // Shraudolphs approximation degrades severely in accuracy, so we return
+    // zero or infinity, depending on the sign of x.
+    return x < 0.0 ? 0.0 : std::numeric_limits<double>::infinity();
```

Comment:
Why not return std::exp(x) ?  This should be a rare case and let it handle overflow

Reply:
@maliberty as mentioned above, it seems to _not_ be a rare case, and significantly hurts performance because `std::exp` is so slow. I think this is something in the root finder that needs to be debugged.

Reply:
Perhaps the root finder keeps chasing denormals values because the stopping criterion is not very good? Capping at -707.7 means that we do not return denormals. (Just a wild guess).

Reply:
The prior code had capping for ```x < -12.0``` so it could well be that the calling code assumes such.   I wonder if you wouldn't get a benefit by keeping the old limit as it was apparently acceptable.   Investigating the solver also makes sense.

I'm guessing >707 doesn't happen much and it could use std::exp.

Reply:
Unfortunately that would yield significantly slower code because we'd need two conditionals. I doubt that the calling code makes such specific assumptions, since the tests pass. With the current version of this PR, you get much better numerics and a 2.1x speedup. What's not to like?

Reply:
Also, if we are OK underflowing at -12, why do we care if we overflow at 707.7 instead of 709.8?